### DISCUSSION TITLE
TypeError: Invalid URL with middleware

### DISCUSSION THREAD

**Discussion by Jeltjo** [vercel/next.js#40344] [Help] [answered]
Wanted to test the new middleware feature in our existing next / react application but with a minimal implementation like below, I'm getting the following error right away, and have no idea what's causing this or how to fix it. 

```
Server Error
TypeError: Invalid URL

This error happened while generating the page. Any console logs will be displayed in the terminal window.
Call Stack
new NodeError
node:internal/errors (372:5)
URL.onParseError
node:internal/url (553:9)
new URL
node:internal/url (629:5)
parseURL
(middleware)/node_modules/next/dist/server/web/next-url.js (181:0)
new NextURL
(middleware)/node_modules/next/dist/server/web/next-url.js (21:0)
adapter
(middleware)/node_modules/next/dist/server/web/adapter.js (15:0)
Module.__WEBPACK_DEFAULT_EXPORT__
(middleware)/node_modules/next/dist/build/webpack/loaders/next-middleware-loader.js?absolutePagePath=%2Fhome%2Fprojects%2Fpwa%2Fmiddleware.ts&page=%2Fmiddleware&matcherRegexp=! (19:47)
<unknown>
file:///home/projects/pwa/node_modules/next/dist/server/web/sandbox/sandbox.js (33:73)
processTicksAndRejections
node:internal/process/task_queues (96:5)
async DevServer.runMiddleware
file:///home/projects/pwa/node_modules/next/dist/server/next-server.js (759:26)

```

middleware.ts:
```
import { NextRequest, NextResponse } from 'next/server';

export function middleware(request: NextRequest) {
	console.log('mw', request.nextUrl);

	return NextResponse.next();
}
```

environment:

    Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP Wed Mar 2 00:30:59 UTC 2022
    Binaries:
      Node: 16.16.0
      npm: 8.11.0
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 12.2.0
      eslint-config-next: 12.2.5
      react: 18.2.0
      react-dom: 18.2.0

Thanks in advance

**Comment by steef17** [comment]
I am also getting an invalid url error. The following error is console logged in the terminal.

`error - (middleware)\node_modules\next\dist\server\web\next-url.js (180:0) @ parseURL
error - Invalid URL: http://undefined:undefined/nl/nl/cart?__nextDefaultLocale=nl&__nextLocale=nl`

What could be the cause of the hostname and port being undefined?

Edit: As seen in the docs (https://nextjs.org/docs/advanced-features/custom-server), the hostname an port is mandatory for the middleware to work. It was working after adding the hostname and port.

**Comment by icyJoseph** [comment, marked as answer]
If both of you are using a custom server, then you need to specify the hostname and port. https://nextjs.org/docs/advanced-features/custom-server:

```tsx
const hostname = 'localhost'
const port = 3000
// when using middleware `hostname` and `port` must be provided below
const app = next({ dev, hostname, port })
const handle = app.getRequestHandler()
```

**Reply by Jeltjo** [reply]
That works! Thanks for the quick reply

**Reply by tokomeno** [reply]
Hi, I'm having the same issue but I use no custom server. Just having the middleware code looks like those

```
import { NextRequest, NextResponse } from 'next/server';

const PUBLIC_FILE = /\.(.*)$/;

export async function middleware(req: NextRequest) {
  if (
    req.nextUrl.pathname.startsWith('/_next') ||
    req.nextUrl.pathname.includes('/api/') ||
    PUBLIC_FILE.test(req.nextUrl.pathname)
  ) {
    return;
  }

  if (req.nextUrl.locale === 'default') {
    return NextResponse.redirect(
      new URL(`/ka${req.nextUrl.pathname}`, req.url)
    );
  }
}
```

**Reply by mkorostoff** [reply]
I am also experiencing this issue while using local middleware, I am also not using a custom server. I am attempting to access the hostname in middleware, but it is always "undefined" (not the primitive [undefined](https://developer.mozilla.org/en-US/docs/Glossary/undefined) but the _string_ "undefined"). As a workaround, I directly inspect the host header:

**Old:**
```javascript
export function middleware(req: NextRequest) {
  const hostname = req.nextUrl.hostname
  console.log(hostname); //logs "undefined" locally
}
```

**New:**
```javascript
export function middleware(req: NextRequest) {
  //Get the host from headers
  const host = req.headers.get('host');
  //Remove the port from the host name, e.g. "localhost:3000" -> "localhost"
  const hostname = host?.split(':')[0];
  console.log(hostname); //logs "localhost" locally
}
```

**Edit:**
Actually! It turned out I was using a custom server, just not one of my own creation. Because my next app runs inside [nx next:serve](https://nx.dev/packages/next/executors/server) I had to specify the hostname and port as shown here: https://github.com/vercel/next.js/discussions/38381#discussioncomment-3094264. Once I did this, I was able to access and update the req.nextUrl.hostname property safely.

**Reply by busaud** [reply]
I applied this fix but it is still not working for me :(

```javascript
const PORT = 443;
const HOSTNAME='example.com';
const dev = process.env.NODE_ENV !== 'production'
const app = next({ dev, HOSTNAME, PORT })
const handle = app.getRequestHandler()
```

```javascript
import { NextResponse } from "next/server";
import type { NextRequest } from "next/server";

export function middleware(request: NextRequest) {
    console.log("Hi");
    console.log(request.url);
    return NextResponse.next();
}

export const config = {
    matcher: "/user"
};
```

result: 
```
error - (middleware)/node_modules/next/dist/esm/server/web/next-url.js (7:0) @ parseURL
error - Invalid URL: https://undefined:undefined/user?__nextDefaultLocale=
null
```
any idea?

Note: I am using the experimental appDir

**Reply by icyJoseph** [reply]
@busaud JavaScript properties are case sensitive:

```js
const PORT = 443;
const HOSTNAME='example.com';
const dev = process.env.NODE_ENV !== 'production'
const app = next({ dev, hostname: HOSTNAME, port: PORT }) // change here
```

**Reply by busaud** [reply]
@icyJoseph Worked! Much appreciated!

**Reply by The-Code-Monkey** [reply]
I'm currently hitting this issue I have middleware but i don't have a custom server, I have an api but its not getting hit no matter what i do.

**Reply by icyJoseph** [reply]
Are you using Nx monorepo?

**Reply by emhat098** [reply]
> @busaud JavaScript properties are case sensitive:
> 
> ```js
> const PORT = 443;
> const HOSTNAME='example.com';
> const dev = process.env.NODE_ENV !== 'production'
> const app = next({ dev, hostname: HOSTNAME, port: PORT }) // change here
> ```

It works for me.
Next.js 13.1.6
Thank you so much.

**Reply by tmoutinho** [reply]
hey @icyJoseph  are you using new appDir? I am trying in nextjs 13 but keep getting this error "Invalid URL". 

I am not using a custom server.

**Reply by chaiyapod** [reply]
> hey @icyJoseph are you using new appDir? I am trying in nextjs 13 but keep getting this error "Invalid URL".
> 
> I am not using a custom server.

i still got like you, I use appDir, Have you solved it yet?

**Reply by icyJoseph** [reply]
If you are getting invalid URL, the issue is most likely that you are using relative paths. On the browser we can do things like `fetch("/some/path")`, and it'll be relative to the document URL, but on a server/edge runtime, there's no document URL, so paths ought to be absolute.

**Reply by chaiyapod** [reply]
> If you are getting invalid URL, the issue is most likely that you are using relative paths. On the browser we can do things like `fetch("/some/path")`, and it'll be relative to the document URL, but on a server/edge runtime, there's no document URL, so paths ought to be absolute.

that works, Thank you so much.

**Comment by mengxuetong** [comment]
async function getList() {
  const res = await fetch(`/api/tools`)
  const data = await res.json()
  return data
}
When I use it on a certain page
The following error has occurred

TypeError: Failed to parse URL from /api/tools
    at new Request (node:internal/deps/undici/undici:9271:19)
    at i (/var/task/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:180562)
    ... 4 lines matching cause stack trace ...
    at async n (/var/task/.next/server/chunks/667.js:1:2187) {
  digest: '448319084',
  [cause]: TypeError: Invalid URL
      at new URL (node:internal/url:797:36)
      at new Request (node:internal/deps/undici/undici:9269:25)
      at i (/var/task/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:180562)
      at G (/var/task/.next/server/chunks/948.js:2:6834)
      at /var/task/.next/server/chunks/948.js:2:9080
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async d (/var/task/.next/server/chunks/667.js:1:2098)
      at async n (/var/task/.next/server/chunks/667.js:1:2187) {
    code: 'ERR_INVALID_URL',
    input: '/api/tools'
  }
}

**Reply by icyJoseph** [reply]
On the server, the URL has to be absolute. Look up why.

In fact, what you are trying to do, is an anti-pattern, https://nextjs-faq.com/fetch-api-in-rsc

**Reply by mengxuetong** [reply]
.env file
NEXT_PUBLIC_URL=https://xxx.vercel.app
Use absolute url:
async function getList() {
  const res = await fetch(`${process.env.NEXT_PUBLIC_URL}/api/tools`)
  const data = await res.json()
  return data
}

it works fine~

**Reply by icyJoseph** [reply]
Right, but it has pitfalls. The right way to do this, is to just get the data from source, no need to do extra API jumps, which are bound to make your app slower, in subtle ways.
